setup.py

```diff
@@ -2,433 +2,517 @@
-    raw_output = subprocess.check_output(
-        [cuda_dir + "/bin/nvcc", "-V"], universal_newlines=True
-    )
-    output = raw_output.split()
-    release_idx = output.index("release") + 1
-    release = output[release_idx].split(".")
-    bare_metal_major = release[0]
-    bare_metal_minor = release[1][0]
-    return (int(bare_metal_major), int(bare_metal_minor))
-
-
-def append_nvcc_threads(nvcc_extra_args):
-    cuda_major, cuda_minor = get_cuda_bare_metal_version(CUDA_HOME)
-    if cuda_major >= 11 and cuda_minor >= 2:
-        return nvcc_extra_args + ["--threads", "4"]
-    return nvcc_extra_args
-
-
-def extra_gencodes(cc_flag):
-    cuda_bare_metal_version = get_cuda_bare_metal_version(CUDA_HOME)
-    if cuda_bare_metal_version >= (11, 0):
-        cc_flag.append("-gencode")
-        cc_flag.append("arch=compute_80,code=sm_80")
-    if cuda_bare_metal_version >= (11, 8):
-        cc_flag.append("-gencode")
-        cc_flag.append("arch=compute_90,code=sm_90")
-
-
-def extra_compiler_flags():
-    extra_flags = [
-        "-O3",
-        "-gencode",
-        "arch=compute_70,code=sm_70",
-        "-U__CUDA_NO_HALF_OPERATORS__",
-        "-U__CUDA_NO_HALF_CONVERSIONS__",
-        "-U__CUDA_NO_BFLOAT16_OPERATORS__",
-        "-U__CUDA_NO_BFLOAT16_CONVERSIONS__",
-        "-U__CUDA_NO_BFLOAT162_OPERATORS__",
-        "-U__CUDA_NO_BFLOAT162_CONVERSIONS__",
-        "-I./transformer_engine/common/layer_norm/",
-        "--expt-relaxed-constexpr",
-        "--expt-extended-lambda",
-        "--use_fast_math",
-    ]
-    if NVTE_WITH_USERBUFFERS:
-        extra_flags.append("-DNVTE_WITH_USERBUFFERS")
-    return extra_flags
-
+from typing import List, Optional, Tuple, Union
 
-cc_flag = []
-extra_gencodes(cc_flag)
+import setuptools
+from setuptools.command.build_ext import build_ext
 
+# Project directory root
+root_path: Path = Path(__file__).resolve().parent
 
-def make_abs_path(l):
-    return [os.path.join(path, p) for p in l]
+@lru_cache(maxsize=1)
+def te_version() -> str:
+    """Transformer Engine version string
 
+    Includes Git commit as local version, unless suppressed with
+    NVTE_NO_LOCAL_VERSION environment variable.
 
-pytorch_sources = [
-    "transformer_engine/pytorch/csrc/extensions.cu",
-    "transformer_engine/pytorch/csrc/common.cu",
-    "transformer_engine/pytorch/csrc/ts_fp8_op.cpp",
-]
-pytorch_sources = make_abs_path(pytorch_sources)
+    """
+    with open(root_path / "VERSION", "r") as f:
+        version = f.readline().strip()
+    if not int(os.getenv("NVTE_NO_LOCAL_VERSION", "0")):
+        try:
+            output = subprocess.run(
+                ["git", "rev-parse" , "--short", "HEAD"],
+                capture_output=True,
+                cwd=root_path,
+                check=True,
+                universal_newlines=True,
+            )
+        except (CalledProcessError, OSError):
+            pass
+        else:
+            commit = output.stdout.strip()
+            version += f"+{commit}"
```

Comment:
Following https://peps.python.org/pep-0440/#local-version-identifiers, everything after `+` is considered a local label. We could also do something like `<major>.<minor>.<patch>+<num commits since release>.<commit>` so that the local labels are ordered, but that gets pretty hairy.

Reply:
Following https://peps.python.org/pep-0440/#local-version-identifiers, everything after `+` is considered a local label. We could also do something like `<major>.<minor>.<patch>+<num commits since release>.<commit>` so that the local labels are ordered, but that gets pretty hairy.